app/src/main/java/org/schabi/newpipe/player/resolver/PlaybackResolver.java

```diff
@@ -67,46 +170,342 @@ default MediaSource buildLiveMediaSource(@NonNull final PlayerDataSource dataSou
-                                         @NonNull final MediaItemTag metadata) {
-        final Uri uri = Uri.parse(sourceUrl);
-        @C.ContentType final int type = TextUtils.isEmpty(overrideExtension)
-                ? Util.inferContentType(uri) : Util.inferContentType("." + overrideExtension);
+    static MediaSource buildMediaSource(@NonNull final PlayerDataSource dataSource,
+                                        @NonNull final Stream stream,
+                                        @NonNull final StreamInfo streamInfo,
+                                        @NonNull final String cacheKey,
+                                        @NonNull final MediaItemTag metadata)
+            throws IOException {
+        if (streamInfo.getService() == ServiceList.YouTube) {
+            return createYoutubeMediaSource(stream, streamInfo, dataSource, cacheKey, metadata);
+        }
 
-        final MediaSource.Factory factory;
-        switch (type) {
-            case C.TYPE_SS:
-                factory = dataSource.getLiveSsMediaSourceFactory();
-                break;
-            case C.TYPE_DASH:
-                factory = dataSource.getDashMediaSourceFactory();
-                break;
-            case C.TYPE_HLS:
-                factory = dataSource.getHlsMediaSourceFactory();
-                break;
-            case C.TYPE_OTHER:
-                factory = dataSource.getExtractorMediaSourceFactory();
-                break;
+        final DeliveryMethod deliveryMethod = stream.getDeliveryMethod();
+        switch (deliveryMethod) {
+            case PROGRESSIVE_HTTP:
+                return buildProgressiveMediaSource(dataSource, stream, cacheKey, metadata);
+            case DASH:
+                return buildDashMediaSource(dataSource, stream, cacheKey, metadata);
+            case HLS:
+                return buildHlsMediaSource(dataSource, stream, cacheKey, metadata);
+            case SS:
+                return buildSSMediaSource(dataSource, stream, cacheKey, metadata);
+            // Torrent streams are not supported by ExoPlayer
             default:
-                throw new IllegalStateException("Unsupported type: " + type);
+                throw new IllegalArgumentException("Unsupported delivery type: " + deliveryMethod);
         }
+    }
 
-        return factory.createMediaSource(
+    @NonNull
+    private static <T extends Stream> ProgressiveMediaSource buildProgressiveMediaSource(
+            @NonNull final PlayerDataSource dataSource,
+            @NonNull final T stream,
+            @NonNull final String cacheKey,
+            @NonNull final MediaItemTag metadata) throws IOException {
+        final String url = stream.getContent();
+
+        if (isNullOrEmpty(url)) {
+            throw new IOException(
+                    "Try to generate a progressive media source from an empty string or from a "
+                            + "null object");
```

Comment:
I agree we should change the exception to another, but I don't think `NullPointerException` is the good one, because the content can be also empty, and so not null. However, I do not see a better `RuntimeException` to throw.

Reply:
I was thinking we could create a custom `ResolveException` extending `Exception`, since I saw that it is then caught somewhere, so it is ok not to use a `RuntimeError`